EIPS/eip-4788.md

```diff
@@ -1,140 +1,109 @@
+and so we only need a constant amount of space to represent this "oracle" in each execution block. To improve the usability of this oracle, block roots are stored
+in a canonical place in the execution state analogous to a `SSTORE` in given contract's storage for each update. Roots are stored keyed by the slot(s) they pertain to.
+To bound the amount of storage this construction consumes, a ring buffer is used that mirrors a block root accumulator on the consensus layer.
+The method for exposing the root data via opcode is inspired by [EIP-2935](./eip-2935.md).
 
 ### Block structure and validity
 
 Beginning at the execution timestamp `FORK_TIMESTAMP`, execution clients **MUST**:
 
-1. set the value of the `ommers` field in the block to an RLP list with one element: the 32 byte [hash tree root](https://github.com/ethereum/consensus-specs/blob/dev/ssz/simple-serialize.md#merkleization) of the [beacon state](https://github.com/ethereum/consensus-specs/blob/dev/specs/bellatrix/beacon-chain.md#beaconstate) from the previous slot to this block.
-
-2. set the value of the `ommersHash` field in the block header to the Keccak256 hash of the `ommers` field.
+1. set the 32 bytes of the execution block header after the `withdrawals_root` to the 32 byte [hash tree root](https://github.com/ethereum/consensus-specs/blob/dev/ssz/simple-serialize.md#merkleization) of the beacon block from the previous slot(s) to this block.
 
-```python
-beaconStateRoot = <32 byte value> # provided by consensus client
-ommers = RLP([beaconStateRoot])   # in the block body
-ommersHash = Keccak256(ommers)    # in the block header
-```
 
-3. Add the block validation that the `ommersHash` does indeed match the expected commitment given the `ommers` value.
+*NOTE*: this field is appended to the current block header structure with this EIP so that the size of the header grows after (and including) the `FORK_TIMESTAMP`.
 
 ### EVM changes
 
 #### Block processing
 
-At the start of processing any execution block where `block.timestamp >= FORK_TIMESTAMP` (i.e. before processing any transactions), write the beacon state root provided in the block into the storage of the smart contract at `HISTORY_STORAGE_ADDRESS`. This data is keyed by the block number.
+At the start of processing any execution block where `block.timestamp >= FORK_TIMESTAMP` (i.e. before processing any transactions), write the beacon root provided in the block header into the storage of the contract at `HISTORY_STORAGE_ADDRESS`. This data is keyed by the slot number.
 
 In pseudocode:
 
 ```python
-beacon_state_root = block.ommers[0]
-sstore(HISTORY_STORAGE_ADDRESS, block.number, beacon_state_root)
+start_timestamp = get_block(block_header.parent_hash).header.timestamp
+start_slot = convert_to_slot(start_timestamp)
+
+end_timestamp = block_header.timestamp
+end_slot = convert_to_slot(end_timestamp)
+
+beacon_block_root = block_header.beacon_block_root
```

Comment:
```suggestion
parent_beacon_root = block_header.parent_beacon_root
```

Reply:
I think this is clear from context but won't die on the naming hill